python/cvmfs/repository.py

```diff
@@ -198,12 +200,170 @@ def _pop_catalog(self):
 
 
 
-class Repository:
-    """ Abstract Wrapper around a CVMFS Repository representation """
+class Cache:
+    def __init__(self, cache_dir, metadata_cleanup):
+        if not os.path.exists(cache_dir):
+            cache_dir = tempfile.mkdtemp(dir='/tmp', prefix='cache.')
+        self._cache_dir = cache_dir
+        self._create_cache_structure()
+        if metadata_cleanup:
+            self._cleanup_metadata()
+
+    def _cleanup_metadata(self):
+        for f in glob.glob(self._cache_dir + '/' + '.cvmfs*'):
+            os.remove(f)
+
+
+    def _create_dir(self, path):
+        cache_full_path = '/'.join([self._cache_dir, path])
+        if not os.path.exists(cache_full_path):
+            os.mkdir(cache_full_path, 0755)
+
+
+    def _create_cache_structure(self):
+        self._create_dir('data')
+        for i in range(0x00, 0xff + 1):
+            new_folder = '{0:#0{1}x}'.format(i, 4)[2:]
+            self._create_dir('data' + '/' + new_folder)
+
+
+    def cache_path(self):
+        return str(self._cache_dir)
+
+
+    def add(self, file_name):
+        full_path = '/'.join([self._cache_dir, file_name])
+        return open(full_path, 'w+')
+
+
+    def get(self, file_name):
+        full_path = '/'.join([self._cache_dir, file_name])
+        if os.path.exists(full_path):
+            return open(full_path, 'rb')
+        return None
+
+
+    def evict(self):
+        for i in range(0x00, 0xff + 1):
+            folder = '{0:#0{1}x}'.format(i, 4)[2:]
+            wildcard = '/'.join([self._cache_dir, folder, '*'])
+            os.remove(glob.glob(wildcard))
+
+
+
+class Fetcher:
+    """ Abstract wrapper around a Fetcher """
+
+    __metadata__ = abc.ABCMeta
+
+    def __init__(self, cache_dir='', metadata_cleanup=False):
+        self.cache = Cache(cache_dir, metadata_cleanup)
+
+
+    @staticmethod
+    def _decompress_file(compressed_file):
+        """ Decompresses a file in a temporary directory """
+        decompressed_data = zlib.decompress(compressed_file.read())
+        temp_file_path = tempfile.mktemp(dir='/tmp')
+        temp_file = open(temp_file_path, 'w+')
+        Fetcher._write_content_into_file(decompressed_data, temp_file)
+        return temp_file
+
+
+    @staticmethod
+    def _write_content_into_file(content, opened_file):
+        """ Writes content into the opened file. The file must have write permission """
+        opened_file.write(content)
+        opened_file.seek(0)
+        opened_file.flush()
+
+
+    @abc.abstractmethod
+    def retrieve_file(self, file_name):
+        """ Abstract method to retrieve a file from the repository """
+        pass
+
+
+
+class LocalFetcher(Fetcher):
+    """ Retrieves files only from the local cache """
+
+    def __init__(self, cache_dir=''):
+        Fetcher.__init__(self, cache_dir, metadata_cleanup=False)
+
+    def retrieve_file(self, file_name):
```

Comment:
We discussed that offline, just for reference: This should expect to access compressed files that are stored locally. For example in the backend storage of a Stratum0 or Stratum1 server.